test-scripting/game_rules.focs.py

```diff
@@ -81,6 +89,7 @@
     default=2,
     min=0,
     max=999,
+    rank=900001,
```

Comment:
Presumably @Cjkjvfnby would prefer not to hard code the ranks like this... Maybe add a Python API getter for a category base rank and current highest ranked rule in a category? Could then calculate ranks for new rules relative to the rules already in the category...

Reply:
problem is mixing python-side and cpp-side categories and trying to put them in arbitrary order also where order on the cpp side is not obvious without explicit rank. 

the weird numbers on ranks I left in may be changed now that I put separate categories. They made it so separate categories were not necessary, but it's moot now. However it still remains an issue how to put cpp-defined game rule between two specific python-defined game rules, or vice versa. 

Think perhaps autogenerating python-side ranks in steps of like 100 would give plenty of space between python game rule definitions and cpp-defined game rules, however then you'd need to calculate the actual ranks "in your head" to specify proper rank on cpp side to put between two python rules... but it's not a very hard calculation if we assume every category on python side starts from 100, then 200 etc.

Reply:
I've implemented what I thought would satisfy various needs, per what I was thinking - have ranking on python side in 100 steps and put internal engine rules in relevant place

Reply:
Feel free to merge it with hardcoded, even without gaps, because we could work on it later, when C++ part is merged.

The idea is to split rules into buckets (range of numbers) and preferably document them.
In Python, each bucket could be a sequence of dynamically generated numbers, defined by order.

So to define an order of the sections and for the C++ rules you still need to hardcode some numbers, but inside the section, it's just an order.   


```python
from itertools import count


# bla-bla-bla 90000 is before ... and after
plant_size_rank = count(90000)



GameRule(
    name="RULE_HABITABLE_SIZE_TINY",
    description="RULE_HABITABLE_SIZE_DESC",
    category="PLANET_SIZE",
    type=int,
    default=1,
    min=0,
    max=999,
    rank=next(plant_size_rank),
)

GameRule(
    name="RULE_HABITABLE_SIZE_MEDIUM",
    description="RULE_HABITABLE_SIZE_DESC",
    category="PLANET_SIZE",
    type=int,
    default=3,
    min=0,
    max=999,
    rank=next(plant_size_rank),
)
```

Reply:
And if we could expose GameRuleRanks  to the python, we could replace `plant_size_rank = count(90000)` with `count(GameRuleRanks.BLA_BLA.value)`

Reply:
When I had to implement a similar ranking in Python , I used a tuple instead of int:  `(ENUM.value, int)`, so first is sorting by the enum value and after that by counter.  In that case, you don't care about real enum value and use order of declaration.

```python
...
(ENUM_SECTION.A, 1),
(ENUM_SECTION.A, 2),
(ENUM_SECTION.B,1),
(ENUM_SECTION.B,2),
...
```